### PR TITLE
[Merged by Bors] - Upgrade to tokio 0.3

### PR DESCRIPTION
**Pull request by AgeManning**
## Description

This PR updates Lighthouse to tokio 0.3. It includes a number of dependency updates and some structural changes as to how we create and spawn tasks.

This also brings with it a number of various improvements:

- Discv5 update
- Libp2p update
- Fix for recompilation issues
- Improved UPnP port mapping handling
- Futures dependency update
- Log downgrade to traces for rejecting peers when we've reached our max

### CODE REVIEW

**Comment by AgeManning** [comment]
This compiles but requires compatibility layers on any dependencies that require tokio 0.2. 

We can add compatibility layers, but the two main deps we need upgraded are warp and rust-libp2p. 

It seems rust-libp2p has a PR pending. I'm going to hold off until we fix a few issues in rust-libp2p and it gets updated to 0.3.

**Comment by AgeManning** [comment]
Without us updating the entire ecosystem, this is now out of our hands, so I'm going to set this to a1

**Comment by AgeManning** [comment]
Thanks to Ben, we're back on for A0

File: beacon_node/eth1/src/service.rs [hunk 1]
```diff
@@ -487,12 +487,12 @@ impl Service {
             .map(|n| n + 1)
             .unwrap_or_else(|| self.config().deposit_contract_deploy_block);
 
-        let range = get_new_block_numbers(
+        let range = tokio_compat_02::FutureExt::compat(get_new_block_numbers(
```
Comment (AgeManning) [inline beacon_node/eth1/src/service.rs:490]: This is likely not necessary, I will remove.

File: beacon_node/eth2_libp2p/Cargo.toml [hunk 2]
```diff
@@ -5,7 +5,8 @@ authors = ["Sigma Prime <contact@sigmaprime.io>"]
 edition = "2018"
 
 [dependencies]
-discv5 = { git = "https://github.com/sigp/discv5", rev = "fba7ceb5cfebd219ebbad6ffdb5d8c31dc8e4bc0", features = ["libp2p"] }
+discv5 = { git = "https://github.com/sigp/discv5", rev = "f117b3ca56fa3dca2317270434634ff7106d391a", features = ["libp2p"] }
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/Cargo.toml:8]: Theres a new tokio 0.3 discv5 I will update too

File: beacon_node/src/lib.rs [hunk 3]
```diff
@@ -76,6 +76,12 @@ impl<E: EthSpec> ProductionBeaconNode<E> {
 
         let executor = context.executor.clone();
 
+        let _guard = executor
```
Comment (AgeManning) [inline beacon_node/src/lib.rs:79]: This should be removed.

**Comment by paulhauner** [comment]
FYI this has some conflicts and a clippy issue.

**Comment by paulhauner** [comment]
Oh also, do you still think we should be targeting `v1.0.0` for this? It seems like there's a decent risk we introduce a critical bug here.

**Comment by paulhauner** [comment]
As per discussions with @AgeManning, I've set this as `blocked` since we don't want to merge it until *after* `v1.0.0`.

**Comment by paulhauner** [comment]
I unblocked this since we've already release `v1.0.0`

**Comment by paulhauner** [comment]
All our Pyrmont nodes are now running the `420` commit, the latest at the time of this comment.

**Pull request closed without merging**